### ISSUE TITLE
tcmalloc cannot optimize std::vector

### ISSUE THREAD

**Issue by NeilZhy** [google/tcmalloc#237]
I run my application with LD_PRELOAD=tcmalloc . My program uses the third-party library [draco](https://github.com/google/draco) when running.The calling code is similar to the sample in draco.
```
draco::DecoderBuffer buffer;
buffer.Init(data.data(), data.size());

const draco::EncodedGeometryType geom_type =
    draco::GetEncodedGeometryType(&buffer);
if (geom_type == draco::TRIANGULAR_MESH) {
  unique_ptr<draco::Mesh> mesh = draco::DecodeMeshFromBuffer(&buffer);
} else if (geom_type == draco::POINT_CLOUD) {
  unique_ptr<draco::PointCloud> pc = draco::DecodePointCloudFromBuffer(&buffer);
}
```
I captured the dynamic memory allocation flame graph using the bcc tool.Combining perf, objdump, gdb and other tools, we found that a hotspot function with high memory usage is inside draco.The final draco code is as follows:
```
bool DataBuffer::Update(const void *data, int64_t size, int64_t offset) {
  if (data == nullptr) {
    if (size + offset < 0)
      return false;
    // If no data is provided, just resize the buffer.
    data_.resize(size + offset);
  } else {
    if (size < 0)
      return false;
    if (size + offset > static_cast<int64_t>(data_.size()))
      data_.resize(size + offset);
    const uint8_t *const byte_data = static_cast<const uint8_t *>(data);
    std::copy(byte_data, byte_data + size, data_.data() + offset);
  }
  descriptor_.buffer_update_count++;
  return true;
}
```
where data_ is defined as std::vector<uint8_t> data_。
Further analysis shows that every time the application calls the draco code, the entire draco will go through multiple nested calls to new to complete the definition of member variables. Therefore, every time data_ is used, it is a newly defined variable, which can actually be understood as a temporary variable.
My question is, why can't tcmalloc optimize away this part of dynamic memory allocation?

**Comment by junyer** [comment]
TCMalloc isn't the C++ compiler. TCMalloc is a memory allocator.

**Comment by NeilZhy** [comment]
> TCMalloc isn't the C++ compiler. TCMalloc is a memory allocator.
Yes, I know TCMalloc is a memory allocator. Please take a closer look at my question. My question is that after using tcmalloc to run the program, there are still a lot of dynamic memory allocations in my program caused by using std::vector.

**Comment by NeilZhy** [comment]
@junyer Yes, I know TCMalloc is a memory allocator. Please take a closer look at my question. My question is that after using tcmalloc to run the program, there are still a lot of dynamic memory allocations in my program caused by using std::vector.

**Comment by junyer** [comment]
Please clarify (a) what a memory allocator should have done, (b) how that could have been done and (c) which memory allocators would have done that. At the risk of stating the obvious, a memory allocator is a library that allocates memory; expecting it to "optimise" how it gets called is like expecting libc to "optimise" how it gets called to, say, `fopen(3)` and `fclose(3)` `FILE`s. There's nothing here beyond wishful thinking.

**Comment by NeilZhy** [comment]
I understand that tcmalloc is a memory pool. When I apply for memory for the first time, tcmalloc caches the memory for me and does not release it to the operating system. What I expect is that after my program runs for a few minutes, there will be no frequent operations of applying for memory from the system. However, after my program runs for a few minutes, I use the bcc tool to capture the flame graph and still find that tcmalloc will apply for physical memory. What I hope to solve is this problem. Of course, we need to apply for memory from tcmalloc, but I don't want tcmalloc to frequently apply for memory from the system. here is the capture picture:
![image](https://github.com/google/tcmalloc/assets/23352727/e6195b2e-3171-4fa0-bcc9-20c19dbd4e5c)

**Comment by thelex0** [comment]
a) these are not tcmalloc function names. are you sure that program using tcmalloc? 
b) tcmalloc has limited cpu cache bounds, so it's not true that there won't be any allocations after a few minutes

**Comment by junyer** [comment]
Indeed, various caches are bounded in various ways. Moreover, it's conceivable that at least some allocations are larger than [`kMaxSize`](https://github.com/google/tcmalloc/blob/85153f3cafacdb64fc26bd2b01ee012d9452e6a0/tcmalloc/common.h#L83).

**Comment by NeilZhy** [comment]
> a) these are not tcmalloc function names. are you sure that program using tcmalloc? b) tcmalloc has limited cpu cache bounds, so it's not true that there won't be any allocations after a few minutes

a. I use tcmalloc to load my program in the following way.
LD_PRELOAD=path/to/gperftools/lib/libtcmalloc.so ./my_test_program     
b. I noticed that github:tcmalloc has the `MallocExtension::instance()->SetMaxPerCpuCacheSize` interface, but these interfaces are not in github:gperftools. I did some configuration by using the github:gperftools's `MallocExtension::instance()->SetNumericProperty(
"tcmalloc.max_total_thread_cache_bytes", 100 * 1024 * 1024);` interface, but it had no effect.

**Comment by junyer** [comment]
Is there some reason why this issue wasn't filed against https://github.com/gperftools/gperftools?

**Comment by NeilZhy** [comment]
I am looking at the issues of github:gperftools one by one, and I haven't found any problems yet.

**Comment by junyer** [comment]
If you aren't using TCMalloc as provided by this project, is there some reason to seek support from this project? gperftools is a separate (older) project with a different (older) implementation of TCMalloc that, in particular, doesn't offer per-CPU caching.

**Comment by NeilZhy** [comment]
Although gperftools uses the old tcmalloc, I hope to get help from tcmalloc. If tcmalloc can definitely solve my problem, I am also happy to use tcmalloc directly.

**Comment by junyer** [comment]
I see you filed https://github.com/gperftools/gperftools/issues/1517, so I think we are done here.

**Comment by NeilZhy** [comment]
Before the problem is solved, I want more people to see the issue, because users of gperftools are not necessarily users of tcmalloc.

**Comment by junyer** [comment]
Spamming unrelated projects is bad open source citizenship, IMO, but good luck with that.

**Comment by NeilZhy** [comment]
I don't want to judge your point of view. I think tcmalloc and gperftools are closely related projects. If this problem is solved, it may help more people.

**Comment by junyer** [comment]
It's also worth noting that even the issue title ("tcmalloc cannot optimize dynamic memory allocation caused by std::vector.") is at best a miscommunication and at worst a misrepresentation. It's unclear that there's any problem here beyond your misunderstanding of how to behave towards open source projects and how to respect people's time and attention.

**Comment by NeilZhy** [comment]
I just stated the facts. I respect everyone and ask questions with an open mind. If you don't have the patience, or you don't have the ability to solve the problem, you can leave it alone. Thank you
